File: 139-word-break/solution.ts
function wordBreak(s: string, wordDict: string[]): boolean {
  const len = s.length;
  const records: boolean[] = [];

  for (let i = 1; i <= len; i++) {
    records[i] = false;
    wordDict.forEach((word) => {
      const start = i - word.length;
      const chunk = s.slice(Math.max(start, 0), i);
      if (chunk === word) {
        records[i] ||= records[start] ?? true;
      }
    });
  }
  return records[len];
}

File: 139-word-break/solution.cpp
class Solution {
public:
    bool wordBreak(string s, vector<string>& wordDict) {
        int n = s.size();
        vector<bool> dp(n + 1, false);
        dp[0] = true;
        for(int i = 1; i <= n; i++) {
            for(auto &w : wordDict) {
                int m = w.size();
                if(i - m >= 0 && s.substr(i - m, m) == w) dp[i] = dp[i] || dp[i - m];
            }
        }
        return dp[n];
    }
};
